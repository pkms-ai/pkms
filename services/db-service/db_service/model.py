from typing import Optional
from enum import Enum
from uuid import UUID

from pydantic import BaseModel


class ContentType(str, Enum):
    WEB_ARTICLE = "web_article"
    PUBLICATION = "publication"
    YOUTUBE_VIDEO = "youtube_video"
    BOOKMARK = "bookmark"
    UNKNOWN = "unknown"


# Pydantic model for updating content
class ContentUpdateModel(BaseModel):
    title: Optional[str] = None
    content_type: Optional[ContentType] = None
    description: Optional[str] = None
    summary: Optional[str] = None
    image_url: Optional[str] = None
    metadata: Optional[dict] = None


# Define the Pydantic model
class ContentModel(BaseModel):
    url: str
    content_type: ContentType
    title: Optional[str] = None
    description: Optional[str] = None
    image_url: Optional[str] = None
    summary: Optional[str] = None
    metadata: Optional[dict] = None
    content_id: Optional[UUID] = None
