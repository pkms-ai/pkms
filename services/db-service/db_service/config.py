import json
from typing import List

from pydantic_settings import BaseSettings


class Settings(BaseSettings):
    POSTGRES_HOST: str = "localhost"
    POSTGRES_PORT: int = 5432
    POSTGRES_USER: str = "postgres"
    POSTGRES_PASSWORD: str
    POSTGRES_DB: str = "content_db"
    CORS_ORIGINS: List[str] = ["http://localhost:3000"]
    ENVIRONMENT: str = "development"
    RETRY_ATTEMPTS: int = 3
    RETRY_DELAY: int = 5  # seconds
    PORT: int = 8003

    def parse_cors_origins(self, v: str | List[str]) -> List[str]:
        """Custom parser for CORS_ORIGINS."""
        if isinstance(v, str):
            try:
                return json.loads(v)
            except json.JSONDecodeError:
                return [v]  # Return the string itself as a single origin if not JSON
        return v  # If it's already a list, return it as is

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        # Parse CORS_ORIGINS if it's a string
        self.CORS_ORIGINS = self.parse_cors_origins(self.CORS_ORIGINS)

    # Use Pydantic v2's model_config instead of Config
    model_config = {"env_file": ".env", "env_file_encoding": "utf-8", "extra": "allow"}

    @property
    def cors_origins(self) -> List[str]:
        return self.CORS_ORIGINS or ["http://localhost:3000"]  # Ensure a default value

    @property
    def database_url(self) -> str:
        return f"postgresql://{self.POSTGRES_USER}:{self.POSTGRES_PASSWORD}@{self.POSTGRES_HOST}:{self.POSTGRES_PORT}/{self.POSTGRES_DB}"


settings = Settings()
