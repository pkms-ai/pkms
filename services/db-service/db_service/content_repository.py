import logging
from enum import Enum
from typing import Optional
from uuid import UUID, uuid4

import asyncpg

logger = logging.getLogger(__name__)


class ContentType(str, Enum):
    WEB_ARTICLE = "web_article"
    YOUTUBE_VIDEO = "youtube_video"
    PUBLICATION = "publication"
    BOOKMARK = "bookmark"
    UNKNOWN = "unknown"


async def check_url_exists(pool: asyncpg.Pool, url: str) -> bool:
    """
    Check if a URL already exists in the database.
    """
    async with pool.acquire() as conn:
        result = await conn.fetchval(
            "SELECT EXISTS(SELECT 1 FROM content WHERE url = $1)", url
        )
    logger.debug(
        f"URL existence check for {url}: {'exists' if result else 'does not exist'}"
    )
    return result


async def get_content_by_id(pool: asyncpg.Pool, content_id: UUID):
    """
    Retrieve content from the database by its ID.
    """
    async with pool.acquire() as conn:
        result = await conn.fetchrow("SELECT * FROM content WHERE id = $1", content_id)
    logger.debug(
        f"Retrieved content for ID {content_id}: {'found' if result else 'not found'}"
    )
    return result


async def get_content_by_url(pool: asyncpg.Pool, url: str):
    """
    Retrieve content from the database by its URL.
    """
    async with pool.acquire() as conn:
        result = await conn.fetchrow("SELECT * FROM content WHERE url = $1", url)
    logger.debug(
        f"Retrieved content for URL {url}: {'found' if result else 'not found'}"
    )
    return result


async def insert_content(
    pool: asyncpg.Pool,
    url: str,
    title: str,
    content_type: ContentType,
    description: Optional[str] = None,
    content_id: Optional[UUID] = None,
) -> UUID:
    """
    Insert new content into the database.
    """
    async with pool.acquire() as conn:
        if content_id is None:
            content_id = uuid4()

        query = """
        INSERT INTO content (id, url, title, content_type, description)
        VALUES ($1, $2, $3, $4::content_type, $5)
        RETURNING id
        """

        inserted_id = await conn.fetchval(
            query, content_id, url, title, content_type.value, description
        )

        logger.info(f"Inserted new content with ID {inserted_id}")
        return inserted_id


async def update_content(
    pool: asyncpg.Pool,
    content_id: UUID,
    title: Optional[str] = None,
    content_type: Optional[ContentType] = None,
    description: Optional[str] = None,
) -> bool:
    """
    Update existing content in the database.
    """
    async with pool.acquire() as conn:
        update_fields = []
        update_values = []
        if title is not None:
            update_fields.append("title = $1")
            update_values.append(title)
        if content_type is not None:
            update_fields.append("content_type = $2::content_type")
            update_values.append(content_type.value)
        if description is not None:
            update_fields.append("description = $3")
            update_values.append(description)

        if not update_fields:
            logger.warning(f"No fields to update for content ID {content_id}")
            return False

        query = f"""
        UPDATE content
        SET {', '.join(update_fields)}, updated_at = CURRENT_TIMESTAMP
        WHERE id = $4
        """
        update_values.append(content_id)

        result = await conn.execute(query, *update_values)
        success = result == "UPDATE 1"
        logger.info(
            f"Updated content with ID {content_id}: {'success' if success else 'failed'}"
        )
        return success


async def delete_content(pool: asyncpg.Pool, content_id: UUID) -> bool:
    """
    Delete content from the database by its ID.
    """
    async with pool.acquire() as conn:
        result = await conn.execute("DELETE FROM content WHERE id = $1", content_id)
        success = result == "DELETE 1"
        logger.info(
            f"Deleted content with ID {content_id}: {'success' if success else 'failed'}"
        )
        return success
