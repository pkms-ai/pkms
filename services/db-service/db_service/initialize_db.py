import asyncio
import asyncpg
from config import settings
import argparse
import uuid

async def table_exists(conn, table_name):
    result = await conn.fetchval(
        """
        SELECT EXISTS (
            SELECT FROM information_schema.tables 
            WHERE table_schema = 'public'
            AND table_name = $1
        )
        """,
        table_name
    )
    return result

async def create_content_table(conn):
    # Create the ContentType enum
    await conn.execute('''
        DO $$ BEGIN
            CREATE TYPE ContentType AS ENUM (
                'web_article', 'youtube_video', 'publication', 'bookmark', 'unknown'
            );
        EXCEPTION
            WHEN duplicate_object THEN null;
        END $$;
    ''')

    # Create the content table with UUID as id and ContentType enum
    await conn.execute('''
        CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
        
        CREATE TABLE IF NOT EXISTS content (
            id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
            url TEXT UNIQUE NOT NULL,
            title TEXT,
            content_type ContentType NOT NULL,
            description TEXT,
            created_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP,
            updated_at TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP
        );
        
        CREATE INDEX IF NOT EXISTS idx_content_url ON content (url);
    ''')
    print("Schema created successfully!")

async def update_schema(conn):
    # Add any schema updates here
    # For example:
    # await conn.execute('''
    #     ALTER TABLE content
    #     ADD COLUMN IF NOT EXISTS new_column TEXT
    # ''')
    print("Schema updated successfully!")

async def populate_content(conn):
    sample_data = [
        (str(uuid.uuid4()), "https://example.com/article1", "Sample Article 1", "web_article"),
        (str(uuid.uuid4()), "https://example.com/article2", "Sample Article 2", "web_article"),
        (str(uuid.uuid4()), "https://youtube.com/watch?v=abc123", "Sample YouTube Video", "youtube_video"),
        (str(uuid.uuid4()), "https://example.com/publication1", "Sample Publication", "publication"),
        (str(uuid.uuid4()), "https://example.com/bookmark1", "Sample Bookmark", "bookmark"),
    ]

    await conn.executemany('''
        INSERT INTO content (id, url, title, content_type)
        VALUES ($1::uuid, $2, $3, $4::ContentType)
        ON CONFLICT (url) DO NOTHING
    ''', sample_data)
    print("Sample data populated successfully!")

async def initialize_database(populate_data=False):
    conn = await asyncpg.connect(settings.database_url)
    try:
        content_table_exists = await table_exists(conn, 'content')
        
        if not content_table_exists:
            await create_content_table(conn)
        else:
            await update_schema(conn)
        
        if populate_data:
            await populate_content(conn)

    finally:
        await conn.close()

def main():
    parser = argparse.ArgumentParser(description="Initialize the database")
    parser.add_argument("--populate-data", action="store_true", help="Populate the database with sample data")

    args = parser.parse_args()

    asyncio.run(initialize_database(populate_data=args.populate_data))

if __name__ == "__main__":
    main()
