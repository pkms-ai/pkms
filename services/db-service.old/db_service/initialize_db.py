import asyncio
import logging
import argparse

from .config import settings

from .db import (
    DatabaseConfig,
    initialize_database,
    create_database,
)

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


async def async_main(host: str):
    db_config = DatabaseConfig(
        host=host,
        port=settings.POSTGRES_PORT,
        user=settings.POSTGRES_USER,
        password=settings.POSTGRES_PASSWORD,
        database=settings.POSTGRES_DB,
    )
    logger.info(f"Attempting to create database with config: {db_config}")
    try:
        await create_database(db_config)
        logger.info(f"Successfully created database {db_config.database}")
        await initialize_database(db_config, populate_sample_data=True)
        logger.info("Database initialization completed successfully")
    except Exception as e:
        logger.error(f"Error during database initialization: {str(e)}", exc_info=True)
        raise


def main():
    parser = argparse.ArgumentParser(description="Initialize the database")
    parser.add_argument("--host", type=str, default=settings.POSTGRES_HOST,
                        help="Database host (overrides the default from settings)")
    args = parser.parse_args()

    asyncio.run(async_main(args.host))


if __name__ == "__main__":
    main()
