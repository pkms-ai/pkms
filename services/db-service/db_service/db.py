import logging

import asyncpg
from pydantic import BaseModel

from db_service.content_repository import ContentType

# Set up logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class DatabaseConfig(BaseModel):
    host: str
    port: int
    user: str
    password: str
    database: str


async def get_db_pool(config: DatabaseConfig):
    """
    Create and return a connection pool for the database.
    """
    logger.info(
        f"Creating database pool for {config.database} at {config.host}:{config.port}"
    )
    return await asyncpg.create_pool(
        host=config.host,
        port=config.port,
        user=config.user,
        password=config.password,
        database=config.database,
    )


async def initialize_database(
    db_config: DatabaseConfig, populate_sample_data: bool = False
):
    pool = await get_db_pool(db_config)
    if not pool:
        logger.error("Error creating database pool")
        return
    try:
        await create_tables_and_indexes(pool)
        logger.info("Database tables and indexes created successfully")

        if populate_sample_data:
            await populate_sample_database(pool)
            logger.info("Database populated with sample data")
    finally:
        await pool.close()


async def create_tables_and_indexes(pool: asyncpg.Pool):
    """
    Initialize the database by creating necessary tables and indexes.
    """
    logger.info("Initializing database...")
    async with pool.acquire() as conn:
        # Create the enum type if it doesn't exist
        await conn.execute("""
            DO $$ BEGIN
                CREATE TYPE content_type AS ENUM (
                    'web_article', 'youtube_video', 'publication', 'bookmark', 'unknown'
                );
            EXCEPTION
                WHEN duplicate_object THEN null;
            END $$;
        """)

        # Create the content table with an index on the url column
        await conn.execute("""
            CREATE TABLE IF NOT EXISTS content (
                id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
                url VARCHAR(2048) UNIQUE NOT NULL,
                content_type content_type NOT NULL,
                title TEXT,
                image_url VARCHAR(2048),
                description TEXT,
                summary TEXT,
                metadata JSONB,
                created_at timestamptz DEFAULT CURRENT_TIMESTAMP,
                updated_at timestamptz DEFAULT CURRENT_TIMESTAMP
            );
            
            CREATE INDEX IF NOT EXISTS idx_content_url ON content (url);
        """)
        # CREATE INDEX IF NOT EXISTS idx_content_title_trgm ON content USING gin (title gin_trgm_ops);
    logger.info("Database initialization complete.")


async def populate_sample_database(pool: asyncpg.Pool):
    """
    Populate the database with sample data.
    """
    sample_data = [
        {
            "url": "https://example.com/article1",
            "title": "Sample Article 1",
            "content_type": ContentType.WEB_ARTICLE.value,
        },
        {
            "url": "https://example.com/article2",
            "title": "Sample Article 2",
            "content_type": ContentType.WEB_ARTICLE.value,
        },
        {
            "url": "https://example.com/video1",
            "title": "Sample Video 1",
            "content_type": ContentType.YOUTUBE_VIDEO.value,
        },
    ]

    try:
        async with pool.acquire() as conn:
            # Insert sample data
            for item in sample_data:
                await conn.execute(
                    """
                    INSERT INTO content (url, title, content_type)
                    VALUES ($1, $2, $3::content_type)
                    ON CONFLICT (url) DO NOTHING
                """,
                    item["url"],
                    item["title"],
                    item["content_type"],
                )

        logger.info(f"Inserted {len(sample_data)} sample items into the database")
    except Exception as e:
        logger.error(f"Error populating database: {str(e)}")
        raise


async def create_database(db_config: DatabaseConfig):
    system_conn = None
    try:
        system_conn = await asyncpg.connect(
            host=db_config.host,
            port=db_config.port,
            user=db_config.user,
            password=db_config.password,
            database="postgres",
        )

        # Check if the database exists
        exists = await system_conn.fetchval(
            "SELECT 1 FROM pg_database WHERE datname = $1", db_config.database
        )

        if not exists:
            # Create the database if it doesn't exist
            await system_conn.execute(f'CREATE DATABASE "{db_config.database}"')
            logger.info(f"Database '{db_config.database}' created successfully")
        else:
            logger.info(f"Database '{db_config.database}' already exists")

    except Exception as e:
        logger.error(f"Error creating database '{db_config.database}': {str(e)}")
        raise
    finally:
        if system_conn:
            await system_conn.close()
