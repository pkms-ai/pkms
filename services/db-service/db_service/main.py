import argparse
import asyncio
import logging
import os
from contextlib import asynccontextmanager
from uuid import UUID

import uvicorn
from fastapi import Depends, FastAPI, HTTPException

from .config import settings
from .content_repository import check_url_exists, insert_content, update_content
from .db import DatabaseConfig, get_db_pool
from .model import (
    ContentModel,
    ContentUpdateModel,
)

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


async def get_db_pool_with_retry(delay: float = 30.0):
    while True:
        try:
            logger.info("Attempting to connect to the database...")
            pool = await get_db_pool(db_config)
            if pool:
                logger.info("Successfully connected to the database")
                return pool
        except Exception as e:
            logger.error(f"Failed to connect to the database: {e}")

        logger.info(f"Retrying in {delay} seconds...")
        await asyncio.sleep(delay)


def get_db_config():
    parser = argparse.ArgumentParser()
    parser.add_argument("--db-host", help="Database host")
    args, _ = parser.parse_known_args()

    db_host = args.db_host or os.environ.get("DB_HOST") or settings.POSTGRES_HOST

    return DatabaseConfig(
        host=db_host,
        port=settings.POSTGRES_PORT,
        user=settings.POSTGRES_USER,
        password=settings.POSTGRES_PASSWORD,
        database=settings.POSTGRES_DB,
    )


# Create DatabaseConfig instance with settings from environment variables
db_config = get_db_config()


@asynccontextmanager
async def lifespan(app: FastAPI):
    # Startup: Create and store the database connection pool
    logger.info("Creating database connection pool")
    app.state.db_pool = await get_db_pool_with_retry()

    yield
    # Shutdown: Close the database connection pool
    if app.state.db_pool:
        logger.info("Closing database connection pool")
        await app.state.db_pool.close()
    else:
        logger.warning("Database connection pool was not initialized")


# Create FastAPI app with lifespan manager
app = FastAPI(lifespan=lifespan)


async def get_db_pool_dependency():
    """
    Dependency to get the database connection pool.
    """
    return app.state.db_pool


@app.get("/check_url")
async def check_url(url: str, pool=Depends(get_db_pool_dependency)):
    """
    Endpoint to check if a URL exists in the database.
    """
    logger.info(f"Checking URL: {url}")
    exists = await check_url_exists(pool, url)
    return {"url": url, "exists": exists}


# Curl command to check /check_url endpoint:
# curl "http://localhost:8003/check_url?url=https://example.com"


# insert conent
@app.post("/content")
async def insert(
    content: ContentModel,
    pool=Depends(get_db_pool_dependency),
):
    """
    Endpoint to insert new content into the database.
    """
    logger.info(f"Inserting new content: {content.url}")

    try:
        inserted_content = await insert_content(pool, content)
        return {"status": "success", "data": inserted_content}
    except ValueError as e:
        logger.error(f"Error inserting content: {str(e)}")
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Unexpected error inserting content: {str(e)}")
        raise HTTPException(status_code=500, detail="Internal Server Error")


#  curl -X POST "http://localhost:8000/content" \
# -H "Content-Type: application/json" \
# -d '{
#     "url": "https://example12.com",
#     "content_type": "web_article",
#     "title": "Sample Title",
#     "description": "This is a sample description.",
#     "summary": "This is a brief summary.",
#     "metadata": {
#         "author": "John Doe",
#         "published_date": "2024-10-21"
#     },
#     "content_id": "f47ac10b-58cc-4372-a567-0e02b2c3d478"
# }'


@app.put("/content/{content_id}")
async def update_content_route(
    content_id: UUID,
    content_update: ContentUpdateModel,
    pool=Depends(get_db_pool_dependency),
):
    """
    Endpoint to update existing content in the database.
    """
    logger.info(f"Updating content with ID: {content_id}")
    try:
        updated_content = await update_content(pool, content_id, content_update)
        if updated_content is None:
            raise HTTPException(
                status_code=404, detail=f"Content with ID {content_id} not found"
            )
        return {"status": "success", "data": updated_content}  # Return as JSON

    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to update content: {str(e)}")
        raise HTTPException(status_code=500, detail="Internal Server Error")


# curl -X PUT "http://localhost:8000/content/f47ac10b-58cc-4372-a567-0e02b2c3d478" \
# -H "Content-Type: application/json" \
# -d '{
#     "title": "Updated Sample Title",
#     "description": "This is an updated sample description.",
#     "summary": "This is an updated brief summary.",
#     "metadata": {
#         "author": "Jane Doe",
#         "published_date": "2024-10-22"
#     }
# }'


@app.get("/health")
async def health_check():
    return {"status": "healthy"}


@app.get("/health/db")
async def health_check_db(pool=Depends(get_db_pool_dependency)):
    try:
        async with pool.acquire() as conn:
            await conn.fetchval(
                "SELECT 1"
            )  # Perform a simple query to check the connection
        return {"status": "healthy"}
    except Exception as e:
        logger.error(f"Database health check failed: {e}")
        raise HTTPException(status_code=503, detail="Database is unavailable")


def start():
    env = settings.ENVIRONMENT
    host = "0.0.0.0"
    port = settings.PORT

    logger.info(f"Starting server in {env} mode on port {port}")
    logger.info(f"CORS origins: {settings.cors_origins}")
    logger.info(f"Database host: {db_config.host}")

    try:
        if env == "development":
            uvicorn.run("db_service.main:app", host=host, port=port, reload=True)
        else:
            uvicorn.run(app, host=host, port=port)
    except Exception as e:
        logger.error(f"Error starting the server: {e}")
    finally:
        logger.info("Shutting down the server")


if __name__ == "__main__":
    start()
