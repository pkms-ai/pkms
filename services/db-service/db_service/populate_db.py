import asyncpg
from db_service.db import ContentType
import logging

logger = logging.getLogger(__name__)

async def populate_database(pool: asyncpg.Pool):
    sample_data = [
        {
            "url": "https://example.com/article1",
            "title": "Sample Article 1",
            "content_type": ContentType.WEB_ARTICLE.value
        },
        {
            "url": "https://example.com/article2",
            "title": "Sample Article 2",
            "content_type": ContentType.WEB_ARTICLE.value
        },
        {
            "url": "https://example.com/video1",
            "title": "Sample Video 1",
            "content_type": ContentType.YOUTUBE_VIDEO.value  # Changed from VIDEO to YOUTUBE_VIDEO
        }
    ]

    try:
        async with pool.acquire() as conn:
            # Insert sample data
            for item in sample_data:
                await conn.execute('''
                    INSERT INTO content (url, title, content_type)
                    VALUES ($1, $2, $3::content_type)
                    ON CONFLICT (url) DO NOTHING
                ''', item['url'], item['title'], item['content_type'])
        
        logger.info(f"Inserted {len(sample_data)} sample items into the database")
    except Exception as e:
        logger.error(f"Error populating database: {str(e)}")
        raise
