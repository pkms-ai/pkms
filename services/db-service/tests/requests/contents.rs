use db_service::{app::App, models::_entities::contents::Entity};
use insta::{assert_debug_snapshot, with_settings};
use loco_rs::testing;
use rstest::rstest;
use sea_orm::EntityTrait;
use serde_json;
use serial_test::serial;

// TODO: see how to dedup / extract this to app-local test utils
// not to framework, because that would require a runtime dep on insta
macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("contents_request");
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_add_content() {
    configure_insta!();

    testing::request::<App, _, _>(|request, _ctx| async move {
        let boot = testing::boot_test::<App>().await.unwrap();
        testing::seed::<App>(&boot.app_context.db).await.unwrap();

        let payload = serde_json::json!({
            "pid": "00000000-0000-0000-0000-000000001234",
            "title": "loco",
            "url": "https://loco.com",
            "content_type": "web_article",
            "image_url": "https://loco.com/image.jpg",
            "description": "loco description",
            "raw_content": "loco raw content",
            "summary": "loco summary",
            "metadata": {
                "key1": "value1",
                "key2": "value2"
            }
        });

        let add_content_request = request.post("/contents").json(&payload).await;

        with_settings!({
            filters => {
                 let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                    combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                    combined_filters
            }
        }, {
            assert_debug_snapshot!(
            (add_content_request.status_code(), add_content_request.text())
        );
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_list() {
    configure_insta!();
    testing::request::<App, _, _>(|request, _ctx| async move {
        let boot = testing::boot_test::<App>().await.unwrap();
        testing::seed::<App>(&boot.app_context.db).await.unwrap();
        let res = request.get("/contents").await;
        // assert_eq!(res.status_code(), 200);

        // you can assert content like this:
        // assert_eq!(res.text(), "content");
        assert_debug_snapshot!((res.status_code(), res.text()));
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_get_content() {
    configure_insta!();

    testing::request::<App, _, _>(|request, _ctx| async move {
        let boot = testing::boot_test::<App>().await.unwrap();
        testing::seed::<App>(&boot.app_context.db).await.unwrap();

        let get_content_request = request.get("/contents/1").await;

        with_settings!({
            filters => {
                let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                combined_filters
            }
        }, {
            assert_debug_snapshot!(
                (get_content_request.status_code(), get_content_request.text())
            );
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_update_content() {
    configure_insta!();

    testing::request::<App, _, _>(|request, _ctx| async move {
        let boot = testing::boot_test::<App>().await.unwrap();
        testing::seed::<App>(&boot.app_context.db).await.unwrap();

        let payload = serde_json::json!({
            "title": "Updated Title",
            "url": "https://updated-url.com",
            "content_type": "publication"
        });

        let update_content_request = request.post("/contents/1").json(&payload).await;

        with_settings!({
            filters => {
                let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                combined_filters
            }
        }, {
            assert_debug_snapshot!(
                (update_content_request.status_code(), update_content_request.text())
            );
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_delete_content() {
    configure_insta!();

    testing::request::<App, _, _>(|request, ctx| async move {
        let boot = testing::boot_test::<App>().await.unwrap();
        testing::seed::<App>(&boot.app_context.db).await.unwrap();

        let count_before_delete = Entity::find().all(&ctx.db).await.unwrap().len();
        let delete_content_request = request.delete("/contents/1").await;

        with_settings!({
            filters => {
                let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                combined_filters
            }
        }, {
            assert_debug_snapshot!(
                (delete_content_request.status_code(), delete_content_request.text())
            );
        });

        let count_after_delete = Entity::find().all(&ctx.db).await.unwrap().len();
        assert_eq!(count_after_delete, count_before_delete - 1);
    })
    .await;
}

#[rstest]
#[case("get_contents", serde_json::json!({}))]
#[case("get_contents_with_page_size", serde_json::json!({"page_size":"1"}))]
#[case("get_contents_with_size_and_page", serde_json::json!({"page":"2", "page_size": "5"}))]
#[case("get_contents_with_filters", serde_json::json!({"page":"1", "page_size": "2", "title": "%loco%"}))]
#[tokio::test]
#[serial]
async fn can_get_contents(#[case] test_name: &str, #[case] params: serde_json::Value) {
    configure_insta!();

    testing::request::<App, _, _>(|request, _ctx| async move {
        let boot = testing::boot_test::<App>().await.unwrap();
        testing::seed::<App>(&boot.app_context.db).await.unwrap();

        let contents = request.get("/contents").add_query_params(params).await;

        with_settings!({
            filters => {
                let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                combined_filters
            }
        }, {
            assert_debug_snapshot!(
                test_name, (contents.status_code(), contents.text())
            );
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn check_url_exists(request: &testing::Request<App>, url: &str) {
    configure_insta!();

    testing::request::<App, _, _>(|request, _ctx| async move {
        let boot = testing::boot_test::<App>().await.unwrap();
        testing::seed::<App>(&boot.app_context.db).await.unwrap();

        let payload = serde_json::json!({
            "url": "https://example.com",
        });

        let update_content_request = request.post("/contents/check_url").json(&payload).await;

        with_settings!({
            filters => {
                let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                combined_filters
            }
        }, {
            assert_debug_snapshot!(
                (update_content_request.status_code(), update_content_request.text())
            );
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn check_url_not_exists(request: &testing::Request<App>, url: &str) {
    configure_insta!();

    testing::request::<App, _, _>(|request, _ctx| async move {
        let boot = testing::boot_test::<App>().await.unwrap();
        testing::seed::<App>(&boot.app_context.db).await.unwrap();

        let payload = serde_json::json!({
            "url": "https://exmple111.com",
        });

        let update_content_request = request.post("/contents/check_url").json(&payload).await;

        with_settings!({
            filters => {
                let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                combined_filters
            }
        }, {
            assert_debug_snapshot!(
                (update_content_request.status_code(), update_content_request.text())
            );
        });
    })
    .await;
}

// add test for empty pid
//
