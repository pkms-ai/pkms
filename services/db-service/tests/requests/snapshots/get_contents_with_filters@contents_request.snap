---
source: tests/requests/contents.rs
expression: "(contents.status_code(), contents.text())"
---
(
    200,
    "[{\"created_at\":\"DATEZ\",\"updated_at\":\"DATEZ\",\"id\":ID,\"pid\":\"b9147c82-f6ac-48a4-8c0a-5d45bc035e76\",\"title\":\"Lorem reptinos\",\"url\":\"https://example.com\",\"content_type\":\"WebArticle\",\"image_url\":\"https://picsum.photos/200\",\"description\":\"This is a simple description.\",\"raw_content\":\"This is raw data.\",\"summary\":\"This is the summary.\",\"metadata\":{\"json\":\"abc\"}},{\"created_at\":\"DATEZ\",\"updated_at\":\"DATEZ\",\"id\":ID,\"pid\":\"aaaa57aa-8af4-46a8-856a-29deebdcd222\",\"title\":\"Another Content\",\"url\":\"https://example.com/article2\",\"content_type\":\"YoutubeVideo\",\"image_url\":\"https://picsum.photos/201\",\"description\":\"Another description.\",\"raw_content\":\"This is another raw data.\",\"summary\":\"Another summary.\",\"metadata\":{\"json\":\"xyz\"}}]",
)
